### PR TITLE
improve the true proposition from the type of equal/eqv/eq

### PR DESCRIPTION
**Pull request by capfredf**
for (equal? a b), when the type of a has an overlap with that of b,
i.e the meet of those two types are not Bottom, put the overlap in the true
proposition.

### CODE REVIEW

File: typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt [hunk 1]
```diff
@@ -63,6 +64,23 @@
      (ret -Boolean (-PS (-is-type o (-val val)) (-not-type o (-val val))))]
     [((tc-result1: (Val-able: (? ok? val))) (tc-result1: t _ o))
      (ret -Boolean (-PS (-is-type o (-val val)) (-not-type o (-val val))))]
+
+    ;; In this case, try to find there is an overlap between t1 and t2
+    [((tc-result1: t1 _ o1) (tc-result1: t2 _ o2))
+     (=> fail)
+
+     (when (ormap (lambda (a) (id=? comparator a))
+                  (list #'member #'memv #'memq))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt:73]: Are these supposed to be `equal?` etc?

Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt:73]: My intention was to not deal with `member` `memv` `memq` for lists in this case, but to handle other binary comparators.

File: typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt [hunk 2]
```diff
@@ -63,6 +64,23 @@
      (ret -Boolean (-PS (-is-type o (-val val)) (-not-type o (-val val))))]
     [((tc-result1: (Val-able: (? ok? val))) (tc-result1: t _ o))
      (ret -Boolean (-PS (-is-type o (-val val)) (-not-type o (-val val))))]
+
+    ;; In this case, try to find there is an overlap between t1 and t2
+    [((tc-result1: t1 _ o1) (tc-result1: t2 _ o2))
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt:69]: use app + #:when

**Review by samth — APPROVED** [review]

File: typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt (hunk 2 above)
Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt:69]: okay, in (app expr pattern), the expr can't reference any variables bound in preceding patterns.

**Pull request merged**